# STEP 3: Importing libraries and Customer dataset

In [1]:
import numpy as np
import pandas as pd
import os

In [2]:
# Setting the path for the datasets
path = r'C:\\Users\\cpaul\\Documents\\Career Foundry\\Instacart Basket Analysis'

# Import customer dataset
df_customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

In [3]:
# Import prepared Instacart data (orders_products_merged_aggregated)
df_orders_products_merged_aggregated = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_aggregated.pkl'))

# STEP 4: Wrangle customer data

In [4]:
df_customers.columns

Index(['user_id', 'First Name', 'Surnam', 'Gender', 'STATE', 'Age',
       'date_joined', 'n_dependants', 'fam_status', 'income'],
      dtype='object')

In [5]:
df_customers.shape

(206209, 10)

In [6]:
df_customers.head(20)

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
0     26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1     33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2     65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3    125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4    130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
5    133128    Cynthia     Noble  Female        Kentucky   43    1/1/2017   
6    152052      Chris    Walton    Male         Montana   20    1/1/2017   
7    168851     Joseph   Hickman    Male  South Carolina   30    1/1/2017   
8     69965     Jeremy      Vang    Male           Texas   47    1/1/2017   
9     82820      Shawn     Chung    Male        Virginia   26    1/1/2017   
10    26441     Gloria  Stafford  Female          Nevada   19    1/1/2017   
11    33835      Roger     Casey    Male        Nebraska   45    1/1/2017   
12    75032    Cynthia  Mcdaniel  Female         Georgia   75    1/1/2017   
13       51  Stephanie    Jensen  Female         Wyoming   37    1/1/2017   
14      516      Peter      Hunt    Male        Colorado   51    1/1/2017   
15     6965    Brandon      Mata    Male    North Dakota   70    1/1/2017   
16     8794        Amy      Mack  Female        Nebraska   30    1/1/2017   
17     9944      Brian    Oneill    Male       Wisconsin   26    1/1/2017   
18    13313     Victor      Moss    Male          Alaska   31    1/1/2017   
19    17335    Patrick    Fowler    Male         Vermont   65    1/1/2017   

    n_dependants                        fam_status  income  
0              3                           married  165665  
1              0                            single   59285  
2              2                           married   99568  
3              0                            single   42049  
4              1                           married   40374  
5              2                           married   49643  
6              0                            single   61746  
7              0                            single   63712  
8              1                           married  162432  
9              2                           married   32072  
10             2  living with parents and siblings   43443  
11             2                           married   47988  
12             1                           married   57835  
13             1                           married   71190  
14             2                           married  146559  
15             0                  divorced/widowed  100091  
16             3                           married   57096  
17             1                           married   91826  
18             2                           married   91111  
19             1                           married   34514

In [7]:
# Renaming columns
df_customers.rename(columns = {'First Name' : 'first_name'}, inplace = True)
df_customers.rename(columns = {'Surnam' : 'last_name'}, inplace = True)
df_customers.rename(columns = {'Gender' : 'gender'}, inplace = True)
df_customers.rename(columns = {'STATE' : 'state'}, inplace = True)
df_customers.rename(columns = {'Age' : 'age'}, inplace = True)
df_customers.rename(columns = {'n_dependants' : 'number_of_dependants'}, inplace = True)
df_customers.rename(columns = {'fam_status' : 'family_status'}, inplace = True)

In [9]:
df_customers.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   number_of_dependants family_status  income  
0                     3       married  165665  
1                     0        single   59285  
2                     2       married   99568  
3                     0        single   42049  
4                     1       married   40374

In [10]:
#Checking descriptive statistics
df_customers.describe()

user_id            age  number_of_dependants         income
count  206209.000000  206209.000000         206209.000000  206209.000000
mean   103105.000000      49.501646              1.499823   94632.852548
std     59527.555167      18.480962              1.118433   42473.786988
min         1.000000      18.000000              0.000000   25903.000000
25%     51553.000000      33.000000              0.000000   59874.000000
50%    103105.000000      49.000000              1.000000   93547.000000
75%    154657.000000      66.000000              3.000000  124244.000000
max    206209.000000      81.000000              3.000000  593901.000000

#  Not planning to drop any columns, these might be useful for the analysis.

# STEP 5: Data quality and consistency checks

In [11]:
#Checking for missing values in the CUSTOMERS dataframe
df_customers.isnull().sum()


user_id                     0
first_name              11259
last_name                   0
gender                      0
state                       0
age                         0
date_joined                 0
number_of_dependants        0
family_status               0
income                      0
dtype: int64

In [12]:
#creating a dataframe for all the Nan values in the first_name field
df_Nan = df_customers[df_customers['first_name'].isnull() == True]

In [14]:
df_Nan.shape

(11259, 10)

In [15]:
df_Nan.head(20)

user_id first_name   last_name  gender           state  age date_joined  \
53     76659        NaN     Gilbert    Male        Colorado   26    1/1/2017   
73     13738        NaN       Frost  Female       Louisiana   39    1/1/2017   
82     89996        NaN      Dawson  Female          Oregon   52    1/1/2017   
99     96166        NaN     Oconnor    Male        Oklahoma   51    1/1/2017   
105    29778        NaN      Dawson  Female            Utah   63    1/1/2017   
128     8562        NaN     Oconnor    Male            Utah   46    1/1/2017   
140   149267        NaN  Hutchinson    Male  South Carolina   20    1/1/2017   
149    82632        NaN         Orr    Male          Hawaii   61    1/1/2017   
155   172331        NaN  Williamson  Female          Alaska   27    1/1/2017   
236   182963        NaN   Nicholson  Female      New Mexico   58    1/2/2017   
241    49727        NaN       Moran    Male          Alaska   53    1/2/2017   
265    63554        NaN  Williamson  Female        Delaware   49    1/2/2017   
275   199686        NaN     Hampton  Female         Montana   21    1/2/2017   
286    36684        NaN    Copeland  Female         Indiana   73    1/2/2017   
294   123632        NaN       Dixon  Female        Delaware   24    1/2/2017   
338    92761        NaN       Floyd  Female       Tennessee   29    1/3/2017   
346   112064        NaN     Dejesus    Male          Kansas   48    1/3/2017   
351   151181        NaN     Oconnor    Male          Kansas   43    1/3/2017   
398    23895        NaN      Patton    Male        New York   56    1/3/2017   
407   169779        NaN     Miranda    Male         Wyoming   32    1/3/2017   

     number_of_dependants family_status  income  
53                      2       married   41709  
73                      0        single   82518  
82                      3       married  117099  
99                      1       married  155673  
105                     3       married  151819  
128                     1       married  134898  
140                     0        single   86778  
149                     1       married  118130  
155                     0        single   55047  
236                     1       married  163391  
241                     1       married  101623  
265                     3       married   30236  
275                     0        single   56294  
286                     2       married  103261  
294                     1       married   36790  
338                     3       married   97830  
346                     2       married   58260  
351                     3       married  128893  
398                     3       married   40763  
407                     2       married   56094

# There are 11259 observations with Nan values for first_name column. I don't want to delete these columns as of now.

In [16]:
# Checking for duplicates
df_customers_dup = df_customers[df_customers.duplicated()]

In [17]:
df_customers_dup

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, number_of_dependants, family_status, income]
Index: []

# No duplicates are found in the customers dataframe

In [18]:
# Checking for mixed-type data

for col in df_customers.columns.tolist():
      weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
      if len (df_customers[weird]) > 0:
            print (col, 'mixed')
      else: print(col, 'consistent')

user_id consistent
first_name mixed
last_name consistent
gender consistent
state consistent
age consistent
date_joined consistent
number_of_dependants consistent
family_status consistent
income consistent


# first_name column has mixed-type data.

In [19]:
# Fix the mixed-type data in the first_name column by changing into a string
df_customers['first_name'] = df_customers['first_name'].astype('str')

In [20]:
# Checking for mixed-type data again to see if it was fixed successfully

for col in df_customers.columns.tolist():
      weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
      if len (df_customers[weird]) > 0:
            print (col, 'mixed')
      else: print(col, 'consistent')

user_id consistent
first_name consistent
last_name consistent
gender consistent
state consistent
age consistent
date_joined consistent
number_of_dependants consistent
family_status consistent
income consistent


In [21]:
#creating a new Nan dataframe for all the Nan values in the first_name field
df_Nan_new = df_customers[df_customers['first_name'].isnull() == True]

In [22]:
df_Nan_new

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, number_of_dependants, family_status, income]
Index: []

# Looks like there are no Nans found in the first_name column after fixing the mixed-type data.

# STEP 6: Combine customer data with the prepared Instacart data

In [23]:
# Checking the datatype of user_id column found in both the dataframes.
df_customers.dtypes

user_id                  int64
first_name              object
last_name               object
gender                  object
state                   object
age                      int64
date_joined             object
number_of_dependants     int64
family_status           object
income                   int64
dtype: object

In [24]:
df_orders_products_merged_aggregated.dtypes

order_id                               int64
user_id                                int64
order_number                           int64
orders_day_of_week                     int64
order_hour_of_day                      int64
days_since_previous_order            float64
product_id                             int64
add_to_cart_order                      int64
reordered                              int64
product_name                          object
aisle_id                               int64
department_id                          int64
prices                               float64
_merge                              category
price_range_loc                       object
busiest_day                           object
busiest_days                          object
busiest_period_of_day                 object
max_order                              int64
loyalty_flag                          object
mean_price                           float64
spender_type_flag                     object
median_day

# The key column, user_id is of type int64 in both dataframes.

In [27]:
#Drop the merge column before the next merge
del df_orders_products_merged_aggregated['_merge']


In [29]:
df_orders_products_merged_aggregated.dtypes

order_id                              int64
user_id                               int64
order_number                          int64
orders_day_of_week                    int64
order_hour_of_day                     int64
days_since_previous_order           float64
product_id                            int64
add_to_cart_order                     int64
reordered                             int64
product_name                         object
aisle_id                              int64
department_id                         int64
prices                              float64
price_range_loc                      object
busiest_day                          object
busiest_days                         object
busiest_period_of_day                object
max_order                             int64
loyalty_flag                         object
mean_price                          float64
spender_type_flag                    object
median_days_since_previous_order    float64
frequency_flag                  

In [30]:
# Now merge the datasets
df_merged = pd.merge(df_orders_products_merged_aggregated, df_customers, on = 'user_id', indicator = True)


In [31]:
df_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_previous_order  product_id  add_to_cart_order  reordered  \
0                        NaN         196                  1          0   
1                       15.0         196                  1          1   
2                       21.0         196                  1          1   
3                       29.0         196                  1          1   
4                       28.0         196                  1          1   

  product_name  ...  first_name  last_name  gender    state age date_joined  \
0         Soda  ...       Linda     Nguyen  Female  Alabama  31   2/17/2019   
1         Soda  ...       Linda     Nguyen  Female  Alabama  31   2/17/2019   
2         Soda  ...       Linda     Nguyen  Female  Alabama  31   2/17/2019   
3         Soda  ...       Linda     Nguyen  Female  Alabama  31   2/17/2019   
4         Soda  ...       Linda     Nguyen  Female  Alabama  31   2/17/2019   

  number_of_dependants  family_status income  _merge  
0                    3        married  40423    both  
1                    3        married  40423    both  
2                    3        married  40423    both  
3                    3        married  40423    both  
4                    3        married  40423    both  

[5 rows x 33 columns]

In [32]:
# Checking the _merge column for full match
df_merged['_merge'].value_counts()

both          32404859
left_only            0
right_only           0
Name: _merge, dtype: int64

In [33]:
#Export merged dataframe in pkl format as instacart_customers_merged.pkl

df_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'instacart_customers_merged.pkl'))
